Mathlib/NumberTheory/ModularForms/Bounds.lean

```diff
@@ -0,0 +1,369 @@
+Authors: David Loeffler
+-/
+module
+
+public import Mathlib.NumberTheory.Modular
+public import Mathlib.NumberTheory.ModularForms.Petersson
+
+/-!
+# Bounds for the norm of a modular form
+
+We prove bounds for the norm of a modular form `f τ` in terms of `im τ`. The main results are
+
+* `ModularFormClass.exists_bound`: a modular form of weight `k` (for a finite-index subgroup `Γ`)
+  is bounded by a constant multiple of `max 1 (1 / (im τ) ^ k))`.
+* `CuspFormClass.exists_bound`: a cusp form of weight `k` (for a finite-index subgroup `Γ`)
+  is bounded by a constant multiple of `1 / (im τ) ^ (k / 2)`.
+
+-/
+@[expose] public section
+
+open Filter Topology Asymptotics Matrix.SpecialLinearGroup Matrix.GeneralLinearGroup
+
+open UpperHalfPlane hiding I
+
+open Matrix hiding mul_smul
+
+open scoped Modular MatrixGroups ComplexConjugate ModularForm
+
+variable {E : Type*} [SeminormedAddCommGroup E]
+
+namespace ModularGroup
+
+/-- The standard fundamental domain truncated at height `y`. -/
+def truncatedFundamentalDomain (y : ℝ) : Set ℍ := { τ | τ ∈ 𝒟 ∧ τ.im ≤ y }
+
+/-- Explicit description of the truncated fundamental domain as a subset of `ℂ`, given by
+obviously closed conditions. -/
+lemma coe_truncatedFundamentalDomain (y : ℝ) :
+    Subtype.val '' truncatedFundamentalDomain y =
+    {z | 0 ≤ z.im ∧ z.im ≤ y ∧ |z.re| ≤ 1 / 2 ∧ 1 ≤ ‖z‖} := by
+  ext z
+  constructor
+  · rintro ⟨⟨z, hz⟩, h, rfl⟩
+    exact ⟨hz.le, h.2, h.1.2, by simpa [Complex.normSq_eq_norm_sq] using h.1.1⟩
+  · rintro ⟨hz, h1, h2, h3⟩
+    have hz' : 0 < z.im := by
+      apply hz.lt_of_ne
+      contrapose! h3
+      simpa [← sq_lt_one_iff₀ (norm_nonneg _), ← Complex.normSq_eq_norm_sq, Complex.normSq,
+        ← h3, ← sq] using h2.trans_lt (by norm_num)
+    exact ⟨⟨z, hz'⟩, ⟨⟨by simpa [Complex.normSq_eq_norm_sq], h2⟩, h1⟩, rfl⟩
+
+/-- For any `y : ℝ`, the standard fundamental domain truncated at height `y` is compact. -/
+lemma isCompact_truncatedFundamentalDomain (y : ℝ) :
+    IsCompact (truncatedFundamentalDomain y) := by
+  rw [Subtype.isCompact_iff, coe_truncatedFundamentalDomain, Metric.isCompact_iff_isClosed_bounded]
+  constructor
+  · -- show closed
+    apply (isClosed_le continuous_const Complex.continuous_im).inter
+    apply (isClosed_le Complex.continuous_im continuous_const).inter
+    apply (isClosed_le (continuous_abs.comp Complex.continuous_re) continuous_const).inter
+    exact isClosed_le continuous_const continuous_norm
+  · -- show bounded
+    refine (Metric.isBounded_iff_subset_closedBall 0).mpr ⟨√((1 / 2) ^ 2 + y ^ 2), fun z hz ↦ ?_⟩
+    simp only [mem_closedBall_zero_iff]
+    refine le_of_sq_le_sq ?_ (by positivity)
+    rw [Real.sq_sqrt (by positivity), Complex.norm_eq_sqrt_sq_add_sq, Real.sq_sqrt (by positivity)]
+    apply add_le_add
+    · rw [sq_le_sq, abs_of_pos <| one_half_pos (α := ℝ)]
+      exact hz.2.2.1
+    · rw [sq_le_sq₀ hz.1 (hz.1.trans hz.2.1)]
+      exact hz.2.1
```

Comment:
I would move this to UpperHalfPlane/Topology, maybe after the verticalStrip stuff? (If the imports work fine for this).

Reply:
Import-bot is complaining, but the increase is pretty small. In that one file it is 3.18%, where anything below 2% is considered noise, but more importantly the downstream import increase is literally zero – every single other file which imports `Modular.lean` also imports `UpperHalfPlane/Topology.lean` – so I think we needn't lose sleep over it.